### PR TITLE
[Fix #8063] Add new `AllowedNames` option for `Naming/ClassAndModuleCamelCase`

### LINKED ISSUE #8063: False positive in Naming/ClassAndModuleCamelCase when using `module_parent`
Naming/ClassAndModuleCamelCase: Use CamelCase for classes and modules. (https://rubystyle.guide#camelcase-classes)

Now that Zeitwerk is out, I prefer defining classes in a single line -- namespace and all.

```ruby
# OLD (Bad)
namespace MyModule
  class MyPrimaryClass
  end
end

# NEW (Good)
class MyModule::MyPrimaryClass
end
```

Sometimes, however, I want to be able to define a minor, sibling class on the current namespace without creating a new file.

```ruby
# Old Way (Bad -- non-collapsed namespace/class)
namespace MyModule
  # Defines: MyModule::MyPrimaryClass (Good)
  class MyPrimaryClass
  end
 
  # Defines: MyModule::MySiblingClass (Good)
  class MySiblingClass
  end
end

# New Way (Good -- collapsed namespace/class)
class MyModule::MyPrimaryClass
  # Bad
  # - Don't want this, because it defines a sub-class, not a sibling class: 
  #   MyModule::MyPrimaryClass::MySiblingClass
  class MySiblingClass
  end

  # Good
  # - Defines the intended sibling class: MyModule::MySiblingClass
  class module_parent::MySiblingClass
  end
end
```

--------

## Expected behavior

The Naming/ClassAndModuleCamelCase cop should allow `module_parent::MySiblingClass` since this is valid syntax that produces the wanted result and since `module_parent` is a class method and is not (nor is intended to be) a constant name.

Perhaps the cop can allow a few keywords that aren't camel-cased as part of a class name? Namely `module_parent`. (I'm not sure if others would be relevant here.)

This fix, if accepted, might add the following to the documentation for this cop:
```ruby
# good
class module_parent::MyClass
end
```

## Actual behavior

Offense raised on this line `class module_parent::MySiblingClass`:
```
Naming/ClassAndModuleCamelCase: Use CamelCase for classes and modules. (https://rubystyle.guide#camelcase-classes)
```

## Steps to reproduce the problem

1. Define:
```ruby
# app/models/my_module/my_primary_class.rb
class MyModule::MyPrimaryClass
  class module_parent::MySiblingClass
  end
end
```
2. Run: `rubocop app/models/my_module/my_primary_class.rb`

## RuboCop version

```
$ [bundle exec] rubocop -V
0.84.0 (using Parser 2.7.1.2, rubocop-ast 0.0.3, running on ruby 2.6.6 x86_64-darwin19)
```

### PR DESCRIPTION
**Pull request by tejasbubane**
Closes #8063

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by bbatsov** [comment]
You should also extend the cop documentation a bit. One thing that concerns me is that it's not very clear that "AllowedNames" can also just a part of a name (the way the code is currently written). I'm not sure I like this, as this adds ambiguity. Better docs or a clearer approach are the way to go here.

File: config/default.yml [hunk 1]
```diff
@@ -2005,6 +2005,9 @@ Naming/ClassAndModuleCamelCase:
   StyleGuide: '#camelcase-classes'
   Enabled: true
   VersionAdded: '0.50'
+  VersionChanged: '0.85'
+  AllowedNames:
```
Comment (bbatsov) [inline config/default.yml:2009]: Options should also be explained here.

**Comment by tejasbubane** [comment]
@bbatsov Added config info to cop documentation & default.yml.

**Comment by bbatsov** [comment]
I see the docs check task is failing. Looking at the diff you made some changes to the cop docs, but didn't run the docs update rake task afterwards.

**Comment by bbatsov** [comment]
I'm afraid you'll have to rebase again. Too much dev activity today. :-)

**Comment by bbatsov** [comment]
Victory! 😆 Thanks for the quick reaction! 🙇‍♂️

**Pull request merged**